README.md
# `bedrock-wp-env`

This is a fork of the WordPress plugin/theme development environment, [`wp-env`](https://github.com/WordPress/gutenberg/tree/trunk/packages/env), which lets you easily set up a local WordPress environment for building and testing plugins and themes. It's simple to install and requires no configuration.

The fork allows you to use the environment for your bedrock setup, where plugins, themes, and WordPress core, are all maintained as part of a version controlled environment.

## Quick (tl;dr) instructions

Ensure that Docker is running, then:

```sh
$ cd /path/to/wordpress/project
$ npm -g i @clorith/bedrock-wp-env
$ bedrock-wp-env start
```

The local environment will be available at http://localhost:8888 (Username: `admin`, Password: `password`).

## Prerequisites

`bedrock-wp-env` requires Docker to be installed. There are instructions available for installing Docker on [Windows 10 Pro](https://docs.docker.com/docker-for-windows/install/), [all other versions of Windows](https://docs.docker.com/toolbox/toolbox_install_windows/), [macOS](https://docs.docker.com/docker-for-mac/install/), and [Linux](https://docs.docker.com/v17.12/install/linux/docker-ce/ubuntu/#install-using-the-convenience-script).

Node.js and NPM are required. The latest LTS version of Node.js is used to develop `wp-env` and is recommended.

Your bedrock installation also needs a `index.php` file in the root directory, this is to prevent additional core files form being downloaded, this is an unfortunate side-effect, as of now, to the Docker images used.

You will want to add the following files to your `.gitignore` file, as they are always automatically generated by the Docker image;

- `./wp-config.php`
- `./phpunit-wp-config.php`

## Installation

### Installation as a global package

After confirming that the prerequisites are installed, you can install `bedrock-wp-env` globally like so:

```sh
$ npm -g i @clorith/bedrock-wp-env
```

You're now ready to use `bedrock-wp-env`!

### Installation as a local package

If your project already has a package.json, it's also possible to use `wp-env` as a local package. First install `wp-env` locally as a dev dependency:

```sh
$ npm i @clorith/bedrock-wp-env --save-dev
```

Then modify your package.json and add an extra command to npm `scripts` (https://docs.npmjs.com/misc/scripts):

```json
"scripts": {
	"bedrock-wp-env": "bedrock-wp-env"
}
```

When installing `bedrock-wp-env` in this way, all `bedrock-wp-env` commands detailed in these docs must be prefixed with `npm run`, for example:

```sh
# You must add another double dash to pass the "update" flag to wp-env
$ npm run bedrock-wp-env start -- --update
```

instead of:

```sh
$ bedrock-wp-env start --update
```

## Usage

### Starting the environment

First, ensure that Docker is running. You can do this by clicking on the Docker icon in the system tray or menu bar.

Then, change to a directory that contains a WordPress project with a Bedrock structure:

```sh
$ cd ~/my-website
```

Then, start the local environment:

```sh
$ bedrock-wp-env start
```

Finally, navigate to http://localhost:8888 in your web browser to see WordPress running with the local WordPress plugin or theme running and activated. Default login credentials are username: `admin` password: `password`.

### Stopping the environment

To stop the local environment:

```sh
$ bedrock-wp-env stop
```

## Troubleshooting common problems

Many common problems can be fixed by running through the following troubleshooting steps in order:

### 1. Check that `bedrock-wp-env` is running

First, check that `bedrock-wp-env` is running. One way to do this is to have Docker print a table with the currently running containers:

```sh
$ docker ps
```

In this table, by default, you should see two entries: `wordpress` with port 8888 and `mariadb` with port 3306.

### 2. Check the port number

By default `bedrock-wp-env` uses port 8888, meaning that the local environment will be available at http://localhost:8888.

You can configure the port that `bedrock-wp-env` uses so that it doesn't clash with another server by specifying the `WP_ENV_PORT` environment variable when starting `bedrock-wp-env`:

```sh
$ WP_ENV_PORT=3333 wp-env start
```

Running `docker ps` and inspecting the `PORTS` column allows you to determine which port `bedrock-wp-env` is currently using.

You may also specify the port numbers in your `.wp-env.json` file, but the environment variables take precedent.

### 3. Restart `bedrock-wp-env`

Restarting `bedrock-wp-env` will restart the underlying Docker containers which can fix many issues.

To restart `bedrock-wp-env`:

```sh
$ bedrock-wp-env stop
$ bedrock-wp-env start
```

### 4. Restart Docker

Restarting Docker will restart the underlying Docker containers and volumes which can fix many issues.

To restart Docker:

1. Click on the Docker icon in the system tray or menu bar.
2. Select _Restart_.

Once restarted, start `bedrock-wp-env` again:

```sh
$ bedrock-wp-env start
```

### 5. Reset the database

Resetting the database which the local environment uses can fix many issues, especially when they are related to the WordPress installation.

To reset the database:

**⚠️ WARNING: This will permanently delete any posts, pages, media, etc. in the local WordPress installation.**

```sh
$ bedrock-wp-env clean all
$ bedrock-wp-env start
```

### 6. Nuke everything and start again 🔥

When all else fails, you can use `bedrock-wp-env destroy` to forcibly remove all of the underlying Docker containers and volumes. This will allow you to start from scratch.

To nuke everything:

**⚠️ WARNING: This will permanently delete any posts, pages, media, etc. in the local WordPress installation.**

```sh
$ bedrock-wp-env destroy
$ bedrock-wp-env start
```

### 7. Debug mode and inspecting the generated dockerfile.

`bedrock-wp-env` uses docker behind the scenes. Inspecting the generated docker-compose file can help to understand what's going on.

Start `bedrock-wp-env` in debug mode

```sh
bedrock-wp-env start --debug
```

`bedrock-wp-env` will output its config which includes `dockerComposeConfigPath`.

```sh
ℹ Config:
	...
	"dockerComposeConfigPath": "/Users/$USERNAME/.wp-env/5a619d332a92377cd89feb339c67b833/docker-compose.yml",
	...
```

## Using Xdebug

Xdebug is installed in the wp-env environment, but it is turned off by default. To enable Xdebug, you can use the `--xdebug` flag with the `wp-env start` command. Here is a reference to how the flag works:

```sh
# Sets the Xdebug mode to "debug" (for step debugging):
bedrock-wp-env start --xdebug

# Sets the Xdebug mode to "off":
bedrock-wp-env start

# Enables each of the Xdebug modes listed:
bedrock-wp-env start --xdebug=profile,trace,debug
```

When you're running `bedrock-wp-env` using `npm run`, like when working in the Gutenberg repo or when having `bedrock-wp-env` as a local project dependency, don't forget to add an extra double dash before the `--xdebug` command:

```sh
bedrock-npm run wp-env start -- --xdebug
```

If you forget about that, the `--xdebug` parameter will be passed to NPM instead of the `wp-env start` command and it will be ignored.

You can see a reference on each of the Xdebug modes and what they do in the [Xdebug documentation](https://xdebug.org/docs/all_settings#mode).

_Since we are only installing Xdebug 3, Xdebug is only supported for PHP versions greater than or equal to 7.2 (the default). Xdebug won't be installed if `phpVersion` is set to a legacy version._

### Xdebug IDE support

To connect to Xdebug from your IDE, you can use these IDE settings. This bit of JSON was tested for VS Code's `launch.json` format (which you can [learn more about here](https://code.visualstudio.com/docs/editor/debugging#_launchjson-attributes)) along with [this PHP Debug extension](https://marketplace.visualstudio.com/items?itemName=felixfbecker.php-debug). Its path mapping also points to a specific plugin -- you should update this to point to the source you are working with inside of the wp-env instance.

You should only have to translate `port` and `pathMappings` to the format used by your own IDE.

```json
{
  "name": "Listen for XDebug",
  "type": "php",
  "request": "launch",
  "port": 9003,
  "pathMappings": {
    "/var/www/html/wp-content/plugins/gutenberg": "${workspaceFolder}/"
  }
}
```

After you create a `.vscode/launch.json` file in your repository, you probably want to add it to your [global gitignore file](https://docs.github.com/en/get-started/getting-started-with-git/ignoring-files#configuring-ignored-files-for-all-repositories-on-your-computer) so that it stays private for you and is not committed to the repository.

Once your IDEs Xdebug settings have been enabled, you should just have to launch the debugger, put a breakpoint on any line of PHP code, and then refresh your browser!

Here is a summary:

1. Start bedrock-wp-env with xdebug enabled: `bedrock-wp-env start --xdebug`
2. Install a suitable Xdebug extension for your IDE if it does not include one already.
3. Configure the IDE debugger to use port `9003` and the correct source files in bedrock-wp-env.
4. Launch the debugger and put a breakpoint on any line of PHP code.
5. Refresh the URL bedrock-wp-env is running at and the breakpoint should trigger.

## Command reference

`bedrock-wp-env` creates generated files in the `bedrock-wp-env` home directory. By default, this is `~/.wp-env`. The exception is Linux, where files are placed at `~/wp-env` [for compatibility with Snap Packages](https://github.com/WordPress/gutenberg/issues/20180#issuecomment-587046325). The `wp-env` home directory contains a subdirectory for each project named `/$md5_of_project_path`. To change the `wp-env` home directory, set the `WP_ENV_HOME` environment variable. For example, running `WP_ENV_HOME="something" wp-env start` will download the project files to the directory `./something/$md5_of_project_path` (relative to the current directory).

### `bedrock-wp-env start`

The start command installs and initializes the WordPress environment, which includes downloading any specified remote sources. By default, `wp-env` will not update or re-configure the environment except when the configuration file changes. Tell `wp-env` to update sources and apply the configuration options again with `bedrock-wp-env start --update`. This will not overwrite any existing content.

```sh
bedrock-wp-env start

Starts WordPress for development on port 8888 (override with WP_ENV_PORT) and
tests on port 8889 (override with WP_ENV_TESTS_PORT). The current working
directory must be a WordPress installation, a plugin, a theme, or contain a
.wp-env.json file. After first install, use the '--update' flag to download
updates to mapped sources and to re-apply WordPress configuration options.

Options:
  --help     Show help                                                 [boolean]
  --version  Show version number                                       [boolean]
  --debug    Enable debug output.                     [boolean] [default: false]
  --update   Download source updates and apply WordPress configuration.
                                                      [boolean] [default: false]
  --xdebug   Enables Xdebug. If not passed, Xdebug is turned off. If no modes
             are set, uses "debug". You may set multiple Xdebug modes by passing
             them in a comma-separated list: `--xdebug=develop,coverage`. See
             https://xdebug.org/docs/all_settings#mode for information about
             Xdebug modes.                                              [string]
```

### `bedrock-wp-env stop`

```sh
bedrock-wp-env stop

Stops running WordPress for development and tests and frees the ports.
```

### `bedrock-wp-env clean [environment]`

```sh
bedrock-wp-env clean [environment]

Cleans the WordPress databases.

Positionals:
  environment  Which environments' databases to clean.
            [string] [choices: "all", "development", "tests"] [default: "tests"]
```

### `bedrock-wp-env run [container] [command]`

The run command can be used to open shell sessions or invoke WP-CLI commands.

<div class="callout callout-alert">
To run a WP-CLI command that includes optional arguments, enclose the WP-CLI command in quotation marks; otherwise, the optional arguments are ignored. This is because flags are normally passed to `wp-env` itself, meaning that the flags are not considered part of the argument that specifies the WP-CLI command. With quotation marks, `wp-env` considers everything inside quotation marks the WP-CLI command argument.

For example, to list cron schedules with optional arguments that specify the fields returned and the format of the output:

```sh
bedrock-wp-env run cli "wp cron schedule list --fields=name --format=csv"
```

Without the quotation marks, WP-CLI lists the schedule in its default format, ignoring the `fields` and `format` arguments.
</div>

Note that quotation marks are not required for a WP-CLI command that excludes optional arguments, although it does not hurt to include them. For example, the following command syntaxes return identical results: `wp-env run cli "wp cron schedule list"` or `wp-env run cli wp cron schedule list`.

For more information about all the available commands, see [WP-CLI Commands](https://developer.wordpress.org/cli/commands/).

```sh
bedrock-wp-env run <container> [command..]

Runs an arbitrary command in one of the underlying Docker containers. The
"container" param should reference one of the underlying Docker services like
"development", or "cli". To run a wp-cli command, use the "cli" service. 
You can also use this command to open shell sessions like
bash and the WordPress shell in the WordPress instance. For example, `bedrock-wp-env run
cli bash` will open bash in the development WordPress instance. When using long
commands with arguments and quotation marks, you need to wrap the "command"
param in quotation marks. For example: `bedrock-wp-env run cli "wp post create
--post_type=page --post_title='Test'"` will create a post on the WordPress
instance.

Positionals:
  container  The container to run the command on.            [string] [required]
  command    The command to run.                           [array] [default: []]

Options:
  --help     Show help                                                 [boolean]
  --version  Show version number                                       [boolean]
  --debug    Enable debug output.                     [boolean] [default: false]
```

For example:

#### Displaying the users on the development instance:

```sh
bedrock-wp-env run cli wp user list
⠏ Running `wp user list` in 'cli'.

ID      user_login      display_name    user_email      user_registered roles
1       admin   admin   wordpress@example.com   2020-03-05 10:45:14     administrator

✔ Ran `wp user list` in 'cli'. (in 2s 374ms)
```

#### Opening the WordPress shell on the development instance and running PHP commands:

```sh
bedrock-wp-env run cli wp shell
ℹ Starting 'wp shell' on the cli container. Exit the WordPress shell with ctrl-c.

Starting 31911d623e75f345e9ed328b9f48cff6_mysql_1 ... done
Starting 31911d623e75f345e9ed328b9f48cff6_tests-wordpress_1 ... done
wp> echo( 'hello world!' );
hello world!
wp> ^C
✔ Ran `wp shell` in 'cli'. (in 16s 400ms)
```

#### Installing a plugin or theme on the development instance

```sh
bedrock-wp-env run cli wp plugin install custom-post-type-ui

Creating 500cd328b649d63e882d5c4695871d04_cli_run ... done
Installing Custom Post Type UI (1.9.2)
Downloading installation package from https://downloads.wordpress.org/plugin/custom-post-type-ui.zip...
The authenticity of custom-post-type-ui.zip could not be verified as no signature was found.
Unpacking the package...
Installing the plugin...
Plugin installed successfully.
Success: Installed 1 of 1 plugins.
✔ Ran `plugin install custom-post-type-ui` in 'cli'. (in 6s 483ms)
```

**NOTE**: Depending on your host OS, you may experience errors when trying to install plugins or themes (e.g. `Warning: Could not create directory.`). This is typically because the user ID used within the container does not have write access to the mounted directories created by `wp-env`. To resolve this, run the `docker-compose` command directly from the directory created by `wp-env` and add `-u $(id -u)` and `-e HOME=/tmp` the `run` command as options:

```sh
$ cd ~/wp-env/500cd328b649d63e882d5c4695871d04
$ docker-compose run --rm -u $(id -u) -e HOME=/tmp cli [plugin|theme] install <plugin|theme>
```

### `wp-env destroy`

```sh
bedrock-wp-env destroy

Destroy the WordPress environment. Deletes docker containers, volumes, and
networks associated with the WordPress environment and removes local files.
```

### `wp-env logs [environment]`

```sh
bedrock-wp-env logs

displays PHP and Docker logs for given WordPress environment.

Positionals:
  environment  Which environment to display the logs from.
      [string] [choices: "development", "tests", "all"] [default: "development"]

Options:
  --help     Show help                                                 [boolean]
  --version  Show version number                                       [boolean]
  --debug    Enable debug output.                     [boolean] [default: false]
  --watch    Watch for logs as they happen.            [boolean] [default: true]
```

### `bedrock-wp-env install-path`

Outputs the absolute path to the WordPress environment files.

Example:

```sh
$ bedrock-wp-env install-path

/home/user/.wp-env/63263e6506becb7b8613b02d42280a49
```

## .wp-env.json

You can customize the WordPress installation, plugins and themes that the development environment will use by specifying a `.wp-env.json` file in the directory that you run `wp-env` from.

`.wp-env.json` supports six fields for options applicable to both the tests and development instances.

| Field                 | Type           | Default                                | Description                                                                                                                               |
|-----------------------|----------------|----------------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------|
| `"phpVersion"`        | `string\null`  | `null`                                 | The PHP version to use. If `null` is specified, `wp-env` will use the default version used with production release of WordPress.          |

CHANGELOG.md
## 4.6.0 (2022-04-03)

### Enhancements
- Automatically implement `.index.php` file in the project root if one does not exist.
- Allow the public directory to be configurable, and automatically detected as `web` or `public` by default.
- Automatically generate a `.htaccess` file in the public directory if none exists, to allow pretty permalinks to work our of the box.

## 4.5.0 (2022-04-02)

### Enhancements
- Added support for the `postSetupCommands` config option, allowing for CLI commands to be triggered on a new site.

## Bug fixes
- Remove default files added by the WordPress Docker image to the project root every time the environment is started.

## 4.2.0 (2022-01-27)

### Enhancement
-   Added command `wp-env install-path` to list the directory used for the environment.
-   The help entry is now shown when no subcommand is passed to `wp-env`.

### Bug Fix
-   Updated `yargs` to fix [CVE-2021-3807](https://nvd.nist.gov/vuln/detail/CVE-2021-3807).

## 4.1.3 (2021-11-07)

### Bug Fix

-   Fix Xdebug installation code to ensure it would fail gracefully

## 4.0.3 (2021-04-29)

### Bug Fix

-   `wp-env destroy` will now work in environments which don't include the `grep` or `awk` commands, such as Windows PowerShell.
-   Fix several permissions issues related to wp-config.php and wp-content files.
-   Fix crash which happened when the path to wp-env's home directory contained a space.
-   Disable Xdebug 3 for PHP versions less than 7.2 to resolve startup crash.

## 4.0.0 (2021-03-17)

### Breaking Change

-   Migrate from `nodegit` to `simple-git`. This change now requires you to have a `git` binary installed locally to utilize the git sources feature of wp-env.json.

### Bug Fixes

-   "mappings" sources are now downloaded if they contain non-local sources.

## 3.0.0 (2020-12-17)

### Breaking Changes

-   `wp-env start` is now the only command which writes to the docker configuration files. Previously, running any command would also parse the config and then write it to the correct location. Now, other commands still parse the config, but they will not overwrite the confugiration which was set by wp-env start. This allows parameters to be passed to wp-env start which can affect the configuration.

### Enhancement

-   Update nodegit dependency to 0.27.0, the earlier version does not have pre-built binaries for Node 14.15.0 LTS. Upgrading provides support without requiring building nodegit locally.
-   Allow WP_HOME wp-config value to be set to a custom port other than the default for the docker instance.
-   Append the instance URL to output of `wp-env start`.

### New feature

-   Add support for setting the PHP version used for the WordPress instance. For example, test PHP 8 with `"phpVersion": 8.0` in wp-env.json.
-   Add Xdebug 3 to the development environment. You can enable Xdebug with `wp-env start --xdebug` (for debug mode) or `wp-env start --xdebug=develop,coverage` for custom modes.

### Bug Fixes

-   ZIP-based plugin sources are now downloaded to a directory using the basename of the URL instead of the full URL path. This prevents HTML encoded characters in the URL (like "/") from being improperly encoded into the filesystem. This fixes the issue where many .zip sources broke because files with these badly formatted characters were not loaded as assets.

## 2.0.0 (2020-09-03)

### Breaking Changes

-   The `config` and `mappings` options in `.wp-env.json` are now merged with any overrides instead of being overwritten.
-   The first listed theme is no longer activated when running wp-env start, since this overwrote whatever theme the user manually activated.
-   `wp-env start` no longer stops the WordPress instance if it was already started unless it needs to configure WordPress.
-   `wp-env start` no longer updates remote sources after first install if the configuration is the same. Use `wp-env start --update` to update sources.

### New Feature

-   You may now specify specific configuration for different environments using `env.tests` or `env.development` in `.wp-env.json`.
-   `wp-env start` is significantly faster after first install.

## 1.6.0-rc.0 (2020-06-24)

### Bug Fixes

-   `wp-env destroy` now removes dangling docker volumes and networks associated with the WordPress environment.

## 1.4.0 (2020-05-28)

### New Feature

-   Add support for running interactive commands. Examples: `wp-env run cli wp shell` and `wp-env run cli bash`.
-   View php and WordPress log output with the new `wp-env logs` command.
-   Clean up your local environment with the new `wp-env destroy` command.
-   Expose Docker service for running phpunit commands.
-   You may now mount local directories to any location within the WordPress install. For example, you may specify `"wp-content/mu-plugins": "./path/to/mu-plugins"` to add mu-plugins.

## 1.1.0 (2020-04-01)

### New Feature

-   URLs for ZIP files are now supported as core, plugin, and theme sources.
-   The `.wp-env.json` coniguration file now accepts a `config` object for setting `wp-config.php` values.
-   A `.wp-env.override.json` configuration file can now be used to override fields from `.wp-env.json`.
-   You may now override the directory in which `wp-env` creates generated files with the `WP_ENV_HOME` environment variable. The default directory is `~/.wp-env/` (or `~/wp-env/` on Linux).
-   The `.wp-env.json` coniguration file now accepts `port` and `testsPort` options which can be used to set the ports on which the docker instance is mounted.

## 1.0.0 (2020-02-10)

### Breaking Changes

-   `wp-env start` no longer accepts a WordPress branch or tag reference as its argument. Instead, create a `.wp-env.json` file and specify a `"core"` field.
-   `wp-env start` will now download WordPress into a hidden directory located in `~/.wp-env`. You may delete your `{projectName}-wordpress` and `{projectName}-tests-wordpress` directories.

### New Feature


lib/config/config.js
'use strict';
/**
 * External dependencies
 */
const fs = require( 'fs' ).promises;
const path = require( 'path' );
const os = require( 'os' );

/**
 * Internal dependencies
 */
const detectDirectoryType = require( './detect-directory-type' );
const { validateConfig, ValidationError } = require( './validate-config' );
const readRawConfigFile = require( './read-raw-config-file' );
const parseConfig = require( './parse-config' );
const md5 = require( '../md5' );
const detectPublicDirectory = require( './detect-public-directory' );

/**
 * wp-env configuration.
 *
 * @typedef WPConfig
 * @property {string}                           name                    Name of the environment.
 * @property {string}                           configDirectoryPath     Path to the .wp-env.json file.
 * @property {string}                           workDirectoryPath       Path to the work directory located in ~/.wp-env.
 * @property {string}                           dockerComposeConfigPath Path to the docker-compose.yml file.
 * @property {boolean}                          detectedLocalConfig     If true, wp-env detected local config and used it.
 * @property {Object.<string, WPServiceConfig>} env                     Specific config for different environments.
 * @property {boolean}                          debug                   True if debug mode is enabled.
 */

/**
 * Base-level config for any particular environment. (development/tests/etc)
 *
 * @typedef WPServiceConfig
 * @property {?WPSource}                 coreSource    The WordPress installation to load in the environment.
 * @property {WPSource[]}                pluginSources Plugins to load in the environment.
 * @property {WPSource[]}                themeSources  Themes to load in the environment.
 * @property {number}                    port          The port to use.
 * @property {Object}                    config        Mapping of wp-config.php constants to their desired values.
 * @property {Object.<string, WPSource>} mappings      Mapping of WordPress directories to local directories which should be mounted.
 * @property {string}                    phpVersion    Version of PHP to use in the environments, of the format 0.0.
 */

/**
 * A WordPress installation, plugin or theme to be loaded into the environment.
 *
 * @typedef WPSource
 * @property {'local'|'git'|'zip'} type     The source type.
 * @property {string}              path     The path to the WordPress installation, plugin or theme.
 * @property {?string}             url      The URL to the source download if the source type is not local.
 * @property {?string}             ref      The git ref for the source if the source type is 'git'.
 * @property {string}              basename Name that identifies the WordPress installation, plugin or theme.
 */

/**
 * Reads, parses, and validates the given .wp-env.json file into a wp-env config
 * object for internal use.
 *
 * @param {string} configPath Path to the .wp-env.json file.
 *
 * @return {WPConfig} A parsed and validated wp-env config object.
 */
module.exports = async function readConfig( configPath ) {
	const configDirectoryPath = path.dirname( configPath );
	const workDirectoryPath = path.resolve(
		await getHomeDirectory(),
		md5( configPath )
	);

	// Default configuration which is overridden by .wp-env.json files.
	const defaultConfiguration = {
		core: null,
		phpVersion: null,
		plugins: [],
		themes: [],
		port: 8888,
		mappings: {},
		postSetupCommands: [],
		publicDirectory: detectPublicDirectory(),
		config: {
			WP_DEBUG: true,
			SCRIPT_DEBUG: true,
			WP_ENVIRONMENT_TYPE: 'local',
			WP_PHP_BINARY: 'php',
			WP_TESTS_EMAIL: 'admin@example.org',
			WP_TESTS_TITLE: 'Test Blog',
			WP_TESTS_DOMAIN: 'http://localhost',
			WP_SITEURL: 'http://localhost',
			WP_HOME: 'http://localhost',
		},
		env: {
			development: {}, // No overrides needed, but it should exist.
			tests: {
				config: { WP_DEBUG: false, SCRIPT_DEBUG: false },
				port: 8889,
			},
		},
	};

	// The specified base configuration from .wp-env.json or from the local
	// source type which was automatically detected.
	const baseConfig =
		( await readRawConfigFile( '.wp-env.json', configPath ) ) ||
		( await getDefaultBaseConfig( configPath ) );

	// Overriden .wp-env.json on a per-user case.
	const overrideConfig =
		( await readRawConfigFile(
			'.wp-env.override.json',
			configPath.replace( /\.wp-env\.json$/, '.wp-env.override.json' )
		) ) || {};

	const detectedLocalConfig =
		Object.keys( { ...baseConfig, ...overrideConfig } ).length > 0;

	// A quick validation before merging on a service by service level allows us
	// to check the root configuration options and provide more helpful errors.
	validateConfig(
		mergeWpServiceConfigs( [
			defaultConfiguration,
			baseConfig,
			overrideConfig,
		] )
	);

	// A unique array of the environments specified in the config options.
	// Needed so that we can override settings per-environment, rather than
	// overwriting each environment key.
	const getEnvKeys = ( config ) => Object.keys( config.env || {} );
	const allEnvs = [
		...new Set( [
			...getEnvKeys( defaultConfiguration ),
			...getEnvKeys( baseConfig ),
			...getEnvKeys( overrideConfig ),
		] ),
	];

	// Returns a pair with the root config options and the specific environment config options.
	const getEnvConfig = ( config, envName ) => [
		config,
		config.env && config.env[ envName ] ? config.env[ envName ] : {},
	];

	// Merge each of the specified environment-level overrides.
	const allPorts = new Set(); // Keep track of unique ports for validation.
	const env = allEnvs.reduce( ( result, environment ) => {
		result[ environment ] = parseConfig(
			validateConfig(
				mergeWpServiceConfigs( [
					...getEnvConfig( defaultConfiguration, environment ),
					...getEnvConfig( baseConfig, environment ),
					...getEnvConfig( overrideConfig, environment ),
				] ),
				environment
			),
			{
				workDirectoryPath,
			}
		);
		allPorts.add( result[ environment ].port );
		return result;
	}, {} );

	if ( allPorts.size !== allEnvs.length ) {
		throw new ValidationError(
			'Invalid .wp-env.json: Each port value must be unique.'
		);
	}

	return withOverrides( {
		name: path.basename( configDirectoryPath ),
		dockerComposeConfigPath: path.resolve(
			workDirectoryPath,
			'docker-compose.yml'
		),
		configDirectoryPath,
		workDirectoryPath,
		detectedLocalConfig,
		env,
	} );
};

/**
 * Deep-merges the values in the given service environment. This allows us to
 * merge the wp-config.php values instead of overwriting them. Note that this
 * merges configs before they have been validated, so the passed config shape
 * will not match the WPServiceConfig type.
 *
 * @param {Object[]} configs Array of raw service config objects to merge.
 *
 * @return {Object} The merged configuration object.
 */
function mergeWpServiceConfigs( configs ) {
	// Returns an array of nested values in the config object. For example,
	// an array of all the wp-config objects.
	const mergeNestedObjs = ( key ) =>
		Object.assign(
			{},
			...configs.map( ( config ) => {
				if ( ! config[ key ] ) {
					return {};
				} else if ( typeof config[ key ] === 'object' ) {
					return config[ key ];
				}
				throw new ValidationError(
					`Invalid .wp-env.json: "${ key }" must be an object.`
				);
			} )
		);

	const mergedConfig = {
		...Object.assign( {}, ...configs ),
		config: mergeNestedObjs( 'config' ),
		mappings: mergeNestedObjs( 'mappings' ),
	};

	delete mergedConfig.env;
	return mergedConfig;
}

/**
 * Detects basic config options to use if the .wp-env.json config file does not
 * exist. For example, if the local directory contains a plugin, that will be
 * added to the default plugin sources.
 *
 * @param {string} configPath A path to the config file for the source to detect.
 * @return {Object} Basic config options for the detected source type. Empty
 *                  object if no config detected.
 */
async function getDefaultBaseConfig( configPath ) {
	const configDirectoryPath = path.dirname( configPath );
	const type = await detectDirectoryType( configDirectoryPath );

	if ( type === 'core' ) {
		return { core: '.' };
	} else if ( type === 'plugin' ) {
		return { plugins: [ '.' ] };
	} else if ( type === 'theme' ) {
		return { themes: [ '.' ] };
	}

	return {};
}

/**
 * Overrides keys in the config object with set environment variables or options
 * which should be merged.
 *
 * @param {WPConfig} config fully parsed configuration object.
 * @return {WPConfig} configuration object with overrides applied.
 */
function withOverrides( config ) {
	// Override port numbers with environment variables.
	config.env.development.port =
		getNumberFromEnvVariable( 'WP_ENV_PORT' ) ||
		config.env.development.port;

	// Override PHP version with environment variable.
	config.env.development.phpVersion =
		process.env.WP_ENV_PHP_VERSION || config.env.development.phpVersion;

	const updateEnvUrl = ( configKey ) => {
		[ 'development' ].forEach( ( envKey ) => {
			try {
				const baseUrl = new URL(
					config.env[ envKey ].config[ configKey ]
				);

				// Don't overwrite the port of WP_HOME when set.
				if ( ! ( configKey === 'WP_HOME' && !! baseUrl.port ) ) {
					baseUrl.port = config.env[ envKey ].port;
				}

				config.env[ envKey ].config[ configKey ] = baseUrl.toString();
			} catch ( error ) {
				throw new ValidationError(
					`Invalid .wp-env.json: config.${ configKey } must be a valid URL.`
				);
			}
		} );
	};

	// Update wp config options to include the correct port number in the URL.
	updateEnvUrl( 'WP_TESTS_DOMAIN' );
	updateEnvUrl( 'WP_SITEURL' );
	updateEnvUrl( 'WP_HOME' );

	return config;
}

/**
 * Parses an environment variable which should be a number.
 *
 * Throws an error if the variable cannot be parsed to a number.
 * Returns null if the environment variable has not been specified.
 *
 * @param {string} varName The environment variable to check (e.g. WP_ENV_PORT).
 * @return {null|number} The number. Null if it does not exist.
 */
function getNumberFromEnvVariable( varName ) {
	// Allow use of the default if it does not exist.
	if ( ! process.env[ varName ] ) {
		return null;
	}

	const maybeNumber = parseInt( process.env[ varName ] );

	// Throw an error if it is not parseable as a number.
	if ( isNaN( maybeNumber ) ) {
		throw new ValidationError(
			`Invalid environment variable: ${ varName } must be a number.`
		);
	}

	return maybeNumber;
}

/**
 * Gets the `wp-env` home directory in which generated files are created.
 *
 * By default: '~/.wp-env/'. On Linux with snap packages: '~/wp-env/'. Can be
 * overriden with the WP_ENV_HOME environment variable.
 *
 * @return {Promise<string>} The absolute path to the `wp-env` home directory.
 */
async function getHomeDirectory() {
	// Allow user to override download location.
	if ( process.env.WP_ENV_HOME ) {
		return path.resolve( process.env.WP_ENV_HOME );
	}

	/**
	 * Installing docker with Snap Packages on Linux is common, but does not
	 * support hidden directories. Therefore we use a public directory when
	 * snap packages exist.
	 *
	 * @see https://github.com/WordPress/gutenberg/issues/20180#issuecomment-587046325
	 */
	return path.resolve(
		os.homedir(),
		!! ( await fs.stat( '/snap' ).catch( () => false ) )
			? 'wp-env'
			: '.wp-env'
	);
}
